### PR TITLE
Bugfix for misaligned splits on hugepages

### PR DESCRIPTION
**Pull request by wfly1998**
There are 5 commits, the first one is to clean up my dirty code 😢 and the second one is a small bugfix for logging on tty/pty.

The last 3 commits is to fix the problem #5463.  The original code did not consider whether the split region was aligned with its page size before splitting, which resulted in an error `VmBoot(MemoryManager(SharedFileSetLen))`.

This PR renames `ram_region_sub_size` to `ram_region_available_size` and make its value align down to the default page size or hugepage size of the current memory zone, which can prevent the memory zone from being split into misaligned parts.  And if the available size of a ram region is zero, this region will be marked as consumed even it has unused space.

Note that there is two methods of using hugepages:

1. Specify `hugepages` in `memory` or `memory-zone` with or without `hugepage_size`.  If the `hugepage_size` is not specified, the value can be got by `statfs` for `/dev/hugepages`
2. Specify a `file` from hugetlbfs, the hugepage size can also be got by `statfs` for it

The value for alignment will be the hugepage size if this memory zone is using hugepages, otherwise the value will be default page size of system.

In addition, the `E820` on `x86_64` and the fdt on `aarch64` does not consider that the front ram region might not be full, this PR also fixes it.

With the following commits, these test cases can pass:

```bash
cloud-hypervisor \
        --memory size=0 \
        --memory-zone id=zone0,size=512M \
        --memory-zone id=zone1,size=3G,hugepages=on,hugepage_size=1G \
        --kernel hypervisor-fw
```

```bash
cloud-hypervisor \
        --memory size=0 \
        --memory-zone id=zone0,size=128M \
        --memory-zone id=zone1,size=1G,hugepages=on,hugepage_size=1G \
        --memory-zone id=zone2,size=128M \
        --memory-zone id=zone3,size=1G,hugepages=on,hugepage_size=1G \
        --memory-zone id=zone4,size=128M \
        --memory-zone id=zone5,size=1G,hugepages=on,hugepage_size=1G \
        --memory-zone id=zone6,size=128M \
        --memory-zone id=zone7,size=1G,hugepages=on,hugepage_size=1G \
        --kernel hypervisor-fw
```

### CODE REVIEW

**Comment by wfly1998** [comment]
@rbradford this problem is more difficult than I imagined 😂 please review it

**Comment by wfly1998** [comment]
Hmm, are these two test cases necessary to be added to the integration test?

File: arch/src/aarch64/fdt.rs [hunk 1]
```diff
@@ -260,32 +260,51 @@ fn create_memory_node(
             fdt.end_node(memory_node)?;
```
Comment (rbradford) [inline arch/src/aarch64/fdt.rs:260]: This commit could do with a little more explanation in the commit message.

**Review by rbradford — APPROVED** [review]
Looks good - especially adding the consideration of the hugetlbfs page size. Your integration test fails are probably due to a missing seccomp entry - probably from `libc::statfs`.

File: arch/src/aarch64/fdt.rs (hunk 1 above)
Comment (wfly1998) [reply arch/src/aarch64/fdt.rs:260]: Ok, necessary explanations are added.

**Comment by wfly1998** [comment]
> Looks good - especially adding the consideration of the hugetlbfs page size. Your integration test fails are probably due to a missing seccomp entry - probably from `libc::statfs`.

Sorry I forgot this, the latest codes are updated.

**Comment by rbradford** [comment]
Still lgtm

File: arch/src/x86_64/mod.rs [hunk 2]
```diff
@@ -827,47 +827,29 @@ pub fn configure_vcpu(
...
+        (layout::RAM_64BIT_START, usize::MAX, RegionType::Ram),
+        // 3 GiB ~ 3712 MiB: 32-bit device memory hole
+        (
+            layout::MEM_32BIT_RESERVED_START,
+            layout::MEM_32BIT_DEVICES_SIZE as usize,
+            RegionType::SubRegion,
+        ),
+        // 3712 MiB ~ 3968 MiB: 32-bit reserved memory hole
+        (
+            layout::MEM_32BIT_RESERVED_START.unchecked_add(layout::MEM_32BIT_DEVICES_SIZE),
```
Comment (likebreath) [inline arch/src/x86_64/mod.rs:848]: Safer to keep using `checked_add` as before, say:
 ```
layout::MEM_32BIT_RESERVED_START
.unchecked_add(layout::MEM_32BIT_DEVICES_SIZE)
.expect("32-bit reserved region is too large")
```

**Review by likebreath — COMMENTED** [review]
I still need to finish the review with the last commit (early next week). 

So far, the only major concern is the way how we create memory mapping from `configure_pvh` is very hard to follow, and sounds error prone to me, as it inexplicitly relies on several assumptions of the guest address space layout. I pushed a new commit to refactor this bits. Please take a look and let me know your thoughts. 

If the new commit makes sense, I will add similar changes for `aarch64`.

**Comment by wfly1998** [comment]
> I still need to finish the review with the last commit (early next week).
> 
> So far, the only major concern is the way how we create memory mapping from `configure_pvh` is very hard to follow, and sounds error prone to me, as it inexplicitly relies on several assumptions of the guest address space layout. I pushed a new commit to refactor this bits. Please take a look and let me know your thoughts.
> 
> If the new commit makes sense, I will add similar changes for `aarch64`.

Looks good, adding those checks is necessary and I ignored it earlier.

Also, I found that I added redundant `)` in the `info` 😂

File: arch/src/x86_64/mod.rs (hunk 2 above)
Comment (wfly1998) [reply arch/src/x86_64/mod.rs:848]: Both `MEM_32BIT_RESERVED_START` and `MEM_32BIT_DEVICES_SIZE` are `GuestAddress(pub u64)` and `checked_add` may not work.  Maybe we can keep using `unchecked_add` but check it by comparing with 4GiB.

**Comment by michael2012z** [comment]
It's surprisingly complex to make the huge pages aligned, especially the handling of the reserved memory hole.

Good fix!

**Comment by michael2012z** [comment]
> I just added the similar refactoring for aarch64, also a new commit to clarify the expected behavior of `create_memory_regions_from_zones`. PTAL. Thanks.

Thank you for covering aarch64 in the refactoring. That looks good.

**Review by michael2012z — APPROVED** [review]

File: arch/src/aarch64/fdt.rs [hunk 3]
```diff
@@ -260,32 +260,86 @@ fn create_memory_node(
...
+                .first()
+                .expect("There should be at last one memory region");
+            let ram_start = super::layout::RAM_START.raw_value();
+            let mem_32bit_reserved_start = super::layout::MEM_32BIT_RESERVED_START.raw_value();
+
+            if !((first_region_start <= &ram_start)
+                && (first_region_end > &ram_start)
+                && (first_region_end <= &mem_32bit_reserved_start))
+            {
+                panic!(
```
Comment (rbradford) [inline arch/src/aarch64/fdt.rs:313]: panic/expect should only be used for programmer error not something that the user can cause. I'm not sure if this is the case in this code but I wanted to highlight that.

Comment (wfly1998) [reply arch/src/aarch64/fdt.rs:313]: Yes, `panic` or `expect` should be used for unexpected errors. And for those cases, neither too few nor too many memory regions should happen.

Also, maybe logging it with `error` and returning `Err` would be better, but there is no suitable `Error` in `FdtWriterResult`.

cc @likebreath

Comment (likebreath) [reply arch/src/aarch64/fdt.rs:313]: Yes, to my understanding, these errors are programming errors, as they indicate the assumptions we reply on are broken (mostly relate to the guest memory layout).

FWIW, I didn't use panic from x86_64 counterpart just to follow the existing manner of how these errors are being handled. To me, these errors are also programming errors.
